### ISSUE TITLE
Decouple processor needs to be added

### ISSUE THREAD

**Issue by mcaparnacoxauto** [aws-observability/aws-otel-lambda#842]
Is your feature request related to a problem? Please describe.

We are currently exploring the possibility of utilizing the Decouple Processor, however, we've noticed that no processors are available as per the information provided on this link: [aws-otel-lambda](https://github.com/aws-observability/aws-otel-lambda#adot-lambda-layer-available-components). This presents a challenge as the Decouple Processor offers unique features that could significantly enhance the performance of our Lambda layer.

Describe the solution you'd like

The Decouple Processor is specifically engineered to circumvent the limitations inherent in Lambda. It does this by reducing overhead and focusing on aspects that are directly relevant to Lambda, making it a more suitable choice compared to other processors. Therefore, it would be highly beneficial to include it in our setup.

One of the key advantages of the Decouple Processor is its awareness of the Lambda [lifecycle](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html#runtimes-extensions-api-lifecycle). This means it can prevent the environment from being frozen or shut down until all pending traces/metrics/logs have been successfully exported. This feature alone could significantly boost the performance of the otel layer.

Furthermore, our Lambda is heavily utilized, processing millions of transactions. Without the Decouple Processor, these transactions could potentially slow down in the ADOT version, affecting overall performance.

Describe alternatives you've considered

We have considered switching to [opentelemetry-lambda](https://github.com/open-telemetry/opentelemetry-lambda/), but this alternative does not include the AWS services that we require.

In conclusion, the inclusion of the Decouple Processor in our setup could provide significant performance improvements, especially considering the high volume of transactions our Lambda processes. We believe this feature request is worth considering to enhance the efficiency and effectiveness of our operations.

**Comment by lukep-coxauto** [comment]
Hello, could we please get some comments by AWS on this issue?  It is blocking us from using this layer because the processor is not in this layer, despite being added to the underlying `opentelemetry-lambda` framework that this layer is built on.

As an alternative, could someone provide clarification on the differences between using this ADOT OTEL layer and using `opentelemetry-lambda` directly?

**Comment by vasireddy99** [comment]
Hello, ADOT Lambda Layer consists a [subset of components](https://github.com/aws-observability/aws-otel-lambda/blob/main/adot/collector/lambdacomponents/default.go#L23-L35) from the upstream and we sometimes apply patches(backport) as need for any fixes. more [information here](https://aws-otel.github.io/docs/getting-started/lambda)

**Comment by lukep-coxauto** [comment]
Hi @vasireddy99  thank you for the response!  

Would you mind providing more direct answers to clarify the AWS team's stance on the following?
1. Will the decouple processor ever be included in the `aws-otel-lambda` project, or is this out of scope for this project?

2. Is there some type of comparison between using `aws-otel-lambda` and `opentelemetry-lambda`?  I cannot tell if one is a superset of the other, or if they are divergent implementations. the README for this project says
> As a downstream Repo of [opentelemetry-lambda](https://github.com/open-telemetry/opentelemetry-lambda)

which indicates to me that this repo is a derived repo based on the previous one.

3. if we choose to use `opentelemetry-lambda` directly, are we losing any integration with Amazon functionality that the `aws-otel-lambda` layer provides? another way to phrase this would be, what is the reason to choose this project over the other one? the README states 
> aws-otel-lambda publishes AWS managed OpenTelemetry Lambda layers that are preconfigured for use with AWS services and bundle the reduced ADOT Collector. Users can onboard to OpenTelemetry in their existing Lambda functions by adding these ready-made layers directly.

but what does that mean for me? do I want the reduced collector? what preconfiguration with AWS services were done?

**Comment by vasireddy99** [comment]
Hi Luke, 

> Will the decouple processor ever be included in the aws-otel-lambda project, or is this out of scope for this project?

We don't enough info about the timeline on when it will be included, cc: @mhausenblas 

> Is there some type of comparison between using aws-otel-lambda and opentelemetry-lambda? I cannot tell if one is a superset of the other, or if they are divergent implementations. the README for this project says

>which indicates to me that this repo is a derived repo based on the previous one.

> ADOT Lambda Layer consists a [subset of components](https://github.com/aws-observability/aws-otel-lambda/blob/main/adot/collector/lambdacomponents/default.go#L23-L35) from the upstream and we sometimes apply patches(backport) as needed for fixes to avoid breaking changes etc., more [information on ADOT Lambda here](https://aws-otel.github.io/docs/getting-started/lambda)

yes we use opentelemetry-lambda repo as a git sub module

> if we choose to use opentelemetry-lambda directly, are we losing any integration with Amazon functionality that the aws-otel-lambda layer provides? 

For the java lambda layers, we made a patch to avoid this [breaking change](https://github.com/open-telemetry/opentelemetry-lambda/issues/714), other than that for the other layers, no integration functionality is lost in general when using otel layers. 

if you are using the SDK layer with in-process exporter flushing the telemetry to the destination, reduced collector may not be necessary.

**Comment by Dr-Emann** [comment]
I believe related to #886 and #787

**Comment by cgilling** [comment]
I just want to add my voice to wanting to get this processor added, seems like it shouldn't be a lot of work to add it since its already a part of the collector. I'm stuck with not really being able to use this since I don't feel comfortable with the added latency that not having the decouple processor adds to user requests, but at the same time I can't use the awsxray trace support with the layers provided by the upstream repo.

**Comment by bhaskarbanerjee** [comment]
@mhausenblas   @vasireddy99 
Please refer to [this](https://aws-observability.github.io/observability-best-practices/guides/serverless/oss/lambda-based-observability-adot/) documentation. It says 
`AWS's managed OpenTelemetry Lambda Layer utilizes OpenTelemetry Lambda Layer to export telemetry data asynchronously from AWS Lambda.`
As far as I understand, ADOT layer is a wrapper on top of open telemetry, which is synchronous in nature as documented [here](https://github.com/open-telemetry/opentelemetry-lambda/blob/main/docs/design_proposal.md#21-lambda-freeze)

The only way to make it async is by including decoupled processor which you are saying is still not part of the ADOT layer, nor is there a roadmap so far.

In that case, I am failing to understand why [aws doc](https://aws-observability.github.io/observability-best-practices/guides/serverless/oss/lambda-based-observability-adot/) is claiming that the process is async.

**Comment by mhausenblas** [comment]
@bhaskarbanerjee thanks for the question and will make sure that we fix the (docs) confusion.

**Comment by fitz-vivodyne** [comment]
still relevant

**Comment by fitz-vivodyne** [comment]
@mhausenblas can you please confirm if the docs are correct and that OTEL information is exported asynchronously, or if @bhaskarbanerjee 's interpretation is correct and it does in fact block the response to the client.

**Comment by bhaskarbanerjee** [comment]
@fitz-vivodyne I have confirmed with @mhausenblas that aws exports data synchronously. It seems the documentation is till not updated @mhausenblas

**Comment by mhausenblas** [comment]
> It seems the documentation is till not update

It was, right after we discussed it, see:

https://github.com/aws-observability/observability-best-practices/commit/470cc2be572bf81f9fce21dca670566db9a68de0

**Comment by bhaskarbanerjee** [comment]
Hey @mhausenblas did that commit get merged and deployed? The page still reads the same.

**Comment by mhausenblas** [comment]
@bhaskarbanerjee thanks again for bringing this to my attention, it's published.

**Comment by fitz-vivodyne** [comment]
Thanks for the follow up @mhausenblas.
I'm assuming that including the decouple processor is still not on the roadmap (as mentioned [here](https://github.com/aws-observability/aws-otel-lambda/issues/842#issuecomment-2049213647))?

If so I'd like to add a +1, the layer would be a lot more compelling if the decouple processor was included and allowed us to trim some of the excess layer latency.

**Comment by jan-xyz** [comment]
Still valid

**Comment by swcloudgenie** [comment]
Still valid for us too

**Comment by pseiferth** [comment]
Still valid here as well

**Comment by Dreamescaper** [comment]
still not stale

**Comment by Dreamescaper** [comment]
Still not stale
